### PR TITLE
Index tag

### PR DESCRIPTION
**Pull request by efirs**
Implements #230

### CODE REVIEW

File: schema/schema_test.go [hunk 1]
```diff
@@ -463,15 +470,30 @@ func TestDefaults(t *testing.T) {
...
+			"type":"object",
+			"default":{"Field1":"aaa"},
+			"properties": {
+				"Field1" : { "type":"string" },
+				"field_nested_required":{"type":"integer"},
+				"field_nested_required2":{"type":"integer"},
+				"field_nested_index2":{"type":"integer"},
+				"field_nested_index_and_req":{"type":"integer"}
+			},
+			"index":["field_nested_index2","field_nested_index_and_req"],
```
Comment (himank) [inline schema/schema_test.go:483]: I'd say have it on the field level in the schema as well. All attributes then will be in the same place like sort, unique, etc.

Comment (efirs) [reply schema/schema_test.go]: makes sense

File: schema/schema_test.go [hunk 2]
```diff
@@ -463,15 +470,28 @@ func TestDefaults(t *testing.T) {
...
 		"field_max_length":{"type":"string","maxLength":123},
 		"field_required":{"type":"integer"},
-		"field_required2":{"type":"integer"}
+		"field_required2":{"type":"integer"},
+		"field_index2":{"type":"integer", "index": true},
+		"field_index_and_req":{"type":"integer", "index": true}
 	},
 	"primary_key":["ID"],
-	"required":["field_required","field_required2"],
+	"required":["field_index_and_req","field_required","field_required2"],
```
Comment (himank) [inline schema/schema_test.go:494]: we should move the required tag as well on the field definition level.

**Review by himank — APPROVED** [review]

**Comment by ovaistariq** [comment]
Do we allow the field to be marked as `required` on the server side? As that is another type of incompatible schema change then in case someone tries to make it optional later.

**Comment by himank** [comment]
No, it is not honored on the server side.

**Comment by ovaistariq** [comment]
> No, it is not honored on the server side. 

Ok great.

**Comment by efirs** [comment]
It is honored on the server side. It's JSON schema standard field.
Incompatible is not making it optional, but adding "required" later.

**Comment by efirs** [comment]
There is a test for it: https://github.com/tigrisdata/tigris/blob/main/test/v1/server/schema_test.go#L215

**Comment by himank** [comment]
Okay, I wasn't sure about it. This needed to be on the field level as well otherwise while processing the schema we don't whether a field is required or not and therefore we won't be doing any backward compatibility checks.

**Comment by efirs** [comment]
It's per JSON schema standard, this is where various tools expect it to be,
for example we don't do anything special, validation works out of the box.
But, you are right, we need to add a compatibility check, if somebody adds it on schema update.

**Comment by ovaistariq** [comment]
> It is honored on the server side. It's JSON schema standard field.
> 
> Incompatible is not making it optional, but adding "required" later.

Yes you are right that's how required fields are specified in json schema: https://json-schema.org/understanding-json-schema/reference/object.html#required-properties

In that case may be index definition could have been the way you originally implemented. 

Logically marking a required field optional is incompatible change from the application perspective as it was previously expecting a field to always be set but after the change there could be a case where the field is not set.

**Comment by himank** [comment]
> It's per JSON schema standard, this is where various tools expect it to be, for example we don't do anything special, validation works out of the box. But, you are right, we need to add a compatibility check, if somebody adds it on schema update.

I am not talking about moving it in the JSON schema. I am talking about ensuring on the server side when we parse the schema we ensure that fields are set with this attribute so that validation can catch this if there is any change on the `required` field list.

**Comment by efirs** [comment]
Yes, we need to make sure it's checked: https://github.com/tigrisdata/tigris/issues/809

**Pull request merged**